modules/core/src/main/java/org/apache/ignite/internal/client/thin/TcpClientTransactions.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.client.thin;
+
+import org.apache.ignite.client.ClientException;
+import org.apache.ignite.client.ClientTransaction;
+import org.apache.ignite.client.ClientTransactions;
+import org.apache.ignite.configuration.ClientTransactionConfiguration;
+import org.apache.ignite.internal.binary.BinaryRawWriterEx;
+import org.apache.ignite.internal.binary.BinaryWriterExImpl;
+import org.apache.ignite.transactions.TransactionConcurrency;
+import org.apache.ignite.transactions.TransactionIsolation;
+
+import static org.apache.ignite.internal.client.thin.ProtocolVersion.V1_5_0;
+
+/**
+ * Implementation of {@link ClientTransactions} over TCP protocol.
+ */
+class TcpClientTransactions implements ClientTransactions {
+    /** Transaction label. */
+    private String lb;
+
+    /** Channel. */
+    private final ReliableChannel ch;
+
+    /** Marshaller. */
+    private final ClientBinaryMarshaller marsh;
+
+    /** Current thread transaction. */
+    private final ThreadLocal<TcpClientTransaction> tx = new ThreadLocal<>();
```

Comment:
It looks like we have a possible memory leak here. `TcpClientTransaction` has a reference to `TcpClientTransactions` which has a reference to `ThreadLocal<TcpClientTransaction> tx`. If some `TcpClientTransactions` are not in use anymore but left a value in thread-local that value and `TcpClientTransactions` instance will never be garbage-collected. Making `TcpClientTransaction` class static should protect from it.

Reply:
We can't make `TcpClientTransaction` static, because we need `TcpClientTransactions` context here. 
I don't quite understand how we can get possible leak here. If nobody holds the reference to `TcpClientTransactions` instance and nobody holds the reference to `TcpClientTransaction` instances, which were created by this `TcpClientTransactions`, the only references left outside of garbage objects it's a `WeakReference` to `ThreadLocal` from threads which have used this `TcpClientTransactions` instance. These weak references don't prevent to collect garbage objects. 
Did I miss something?

Reply:
Sorry, I got it. There is a weak reference only to `ThreadLocal` instance, but regular reference to `TcpClientTransaction` from the thread-local table, so memory leak is possible.
I will try to fix it tomorrow.

Reply:
I've fixed the implementation. Also, I've found and fixed server-side transactions leak (transaction has left in active set after an attempt to end it when it's impossible to resume the transaction).

Reply:
👍